### PR TITLE
[RFC] Dagster Celery Runner

### PR DESCRIPTION
**Pull request by egordm**
# Summary & Motivation

Hello, I would like to propose a new feature to Dagster celery library by introducing Dagster Celery Runner. The PR also includes changes described in this RFC. We have been running this code in production for a while already and are confident that it is polished enough to be considered to be included as part of Dagster. 

## Context and scope
Our workload consists of:
* Multiple assets, each with on average 300 partitions.
* Hourly schedule to materialize these assets (for each partition).
* Each materialization job takes between 10 and 50 seconds.

Using Dagster K8s run launcher along with single process executor, we have noticed that:
* The workload puts too much load on our K8s cluster.
* Startup times become significant due to the small size of our jobs.
Without overcomplicating the solution, this setup did not seem as the way to go.

K8s Runner drawio

## Goals and non-goals
Our goal was to create setup which:
* Supports our workload.
* Minimizes the run launching overhead on our infrastructure.
* Maintains low job startup latency.
* At-least-once execution strategy (runners like k8s give us)
* Ability to scale horizontally.
* Run level isolation.
* Minimize downtime during deployment.

Non-goals:
* Op level isolation

### Out-of-the-Box Dagster Celery
Original [dagster-celery](https://docs.dagster.io/deployment/guides/celery) package provides an alternative method of launching runs by dividing op execution across different workers.

This approach still relies on a runner to start the run, within which the executor passes the tasks onto individual celery workers. Using it together with K8s executor would not solve our problem, and combining it with Default Runner (starts runs on using the code server) would introduce new scalability and deployment issues (see diagram).

Celery Executor drawio

While measuring performance of Dagster Celery executor, we did conclude that it did have the desired run startup latency and potential scalability.

## The actual design
Our proposed approach introduces a Dagster Celery runner which mimics the way K8s runner works, but starts the runs directly on Dagster celery workers (see diagram).

Celery Runner drawio

The main difference in this approach is that instead of launching a process containing an executor to monitor the steps, the task is submitted directly to the Celery Broker which in turn finds a worker to start the execution.

Once a task has been submitted to celery the daemon monitors the process using standard job monitoring requests which run at a configured interval. From celery backend the monitoring can read whether a job is 'in progress', 'done' or an alternative state.

### At-least-once-execution
Our production setup runs workers on k8s and uses stateless Redis as both Celery Broker and Backend.

There are three edge cases when using Celery with stateless Broker and Backend:
1. Redis broker crashes midway before starting the job, causing queue to drop jobs before they are picked up.
2. Redis backend crashes before the success result is communicated.
3. The worker is forcibly terminated, thus not communicating the result to the celery backend.

We mitigate case (1) by setting a sensible `start_timeout_seconds` on for run monitoring. If a job is not started in time, it will be marked as failed and caught by the retry policy.

Case (2) is handled out of the box as the worker still communicates the job status to Dagster run storage. In case a job is marked as completed in Dagster, the monitoring stops.

To handle case (3) we enforce a mandatory `dagster/max_runtime` tag on all jobs as it is otherwise impossible to distinguish a very slow job from a terminated one.

### Scalability
Because the workers facilitate job execution, scaling is as easy as adding more workers. The job monitoring is cheap because it requires a call to Redis backend from the monitoring daemon.

To make the solution more flexible, we add a run level tag specifying the celery queue to start the job on, thus allowing different workers to process different runs (akin to current op level queue tag).

### Deployment
Celery supports warm shutdown, meaning a worker will stop accepting new tasks once it receives a SIGTERM signal, but will finish existing ones. We combine this feature with k8s termination grace period and default rolling update to spin up new workers to pick up the new tasks.

### Run isolation
Using celery instead of k8s loses the resource isolation jobs have. We use default prefork pool along with `max-memory-per-child` and `max-tasks-per-child` settings to introduce soft limits on worker processes.

# How I Tested These Changes
* Created new tests based on the default runner tests. 
* Created example docker-compose.

### CODE REVIEW

**Comment by egordm** [comment]
Hi @alangenfeld . I was wondering if someone could take a look or provide a review for this PR?
Thanks in advance.

**Review by alangenfeld — CHANGES_REQUESTED** [review]
Overall this seems reasonable to me, let me double check with the team about any concerns there would be about bring this in and maintaining it over time. 

Lets get the build passing by addressing the issues in https://buildkite.com/dagster/unit-tests/builds/740

File: python_modules/libraries/dagster-celery/dagster_celery_tests/repo_runner.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+
+@job(tags=DEFAULT_TAGS)
+def math_diamond():
+    one = return_one()
+    add(multiply_by_2(one), multiply_by_3(one))
+
+
+@repository
+def nope():
```
[the remark this answers is not in the document]
Comment (egordm) [reply python_modules/libraries/dagster-celery/dagster_celery_tests/repo_runner.py]: That's bad naming indeed. :sweat_smile: 
I have updated it to a more descriptive name "celery_test_repository"

File: python_modules/libraries/dagster-celery/dagster_celery/make_app.py [hunk 2]
```diff
@@ -1,17 +1,38 @@
...
+                "queue": "dagster",
+                "routing_key": f"dagster.{TASK_EXECUTE_PLAN_NAME}",
+            },
+            TASK_EXECUTE_JOB_NAME: {
+                "queue": "dagster",
+                "routing_key": f"dagster.{TASK_EXECUTE_PLAN_NAME}",
+            },
+            TASK_RESUME_JOB_NAME: {
+                "queue": "dagster",
+                "routing_key": f"dagster.{TASK_EXECUTE_PLAN_NAME}",
```
[the remark this answers is not in the document]
Comment (egordm) [reply python_modules/libraries/dagster-celery/dagster_celery/make_app.py]: Good find! The keys should indeed align. Fixed!

File: python_modules/libraries/dagster-celery/dagster_celery/launcher.py [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+    celery: Celery
+
+    def __init__(
+        self,
+        default_queue: str,
+        broker: str | None = None,
+        backend: str | None = None,
+        include: list[str] | None = None,
+        config_source: dict | None = None,
+        inst_data: ConfigurableClassData | None = None,
```
[the remark this answers is not in the document]
Comment (egordm) [reply python_modules/libraries/dagster-celery/dagster_celery/launcher.py]: I have fixed CI build errors. The 3.10 union naming is now replaced by the `Optional` type.

**Comment by egordm** [comment]
Thanks for looking into it! Sounds like a good plan. Let me know if there are any other changes necessary.

**Review by alangenfeld — APPROVED** [review]
While there are still concerns over the long term maintenance of this, this is a really well put together PR and

> We have been running this code in production for a while already and are confident

helps me believe that this should be usable and useful to others even if its slow to receive any further attention from the Dagster team directly. 

Thanks for the contribution!

**Comment by egordm** [comment]
Awesome! Thanks!

**Pull request merged**